repose-aggregator/components/filters/cors/src/main/scala/org/openrepose/filters/cors/CorsFilter.scala

```diff
@@ -0,0 +1,180 @@
+import org.openrepose.commons.utils.http.{HeaderConstant, CommonHttpHeader}
+import org.openrepose.core.filter.FilterConfigHelper
+import org.openrepose.core.services.config.ConfigurationService
+import org.openrepose.filters.cors.config.CorsConfig
+
+import scala.collection.JavaConverters._
+import scala.language.implicitConversions
+import scala.util.matching.Regex
+
+@Named
+class CorsFilter @Inject()(configurationService: ConfigurationService)
+  extends Filter with UpdateListener[CorsConfig] with LazyLogging {
+
+  import CorsFilter._
+
+  private var configurationFile: String = DEFAULT_CONFIG
+  private var initialized = false
+  private var allowedOrigins: Iterable[Regex] = _
+  private var allowedMethods: Iterable[String] = _
+  private var resources: Iterable[Resource] = _
+
+  override def init(filterConfig: FilterConfig): Unit = {
+    logger.trace("CORS filter initializing...")
+    configurationFile = new FilterConfigHelper(filterConfig).getFilterConfig(DEFAULT_CONFIG)
+
+    logger.info(s"Initializing CORS Filter using config $configurationFile")
+    val xsdUrl: URL = getClass.getResource(SCHEMA_FILE_NAME)
+    configurationService.subscribeTo(filterConfig.getFilterName, configurationFile, xsdUrl, this, classOf[CorsConfig])
+
+    logger.trace("CORS filter initialized.")
+  }
+
+  override def doFilter(servletRequest: ServletRequest, servletResponse: ServletResponse, filterChain: FilterChain): Unit = {
+    val httpServletRequest = servletRequest.asInstanceOf[HttpServletRequest]
+    val httpServletResponse = servletResponse.asInstanceOf[HttpServletResponse]
+    val isOptions = httpServletRequest.getMethod == "OPTIONS"
+    val origin = httpServletRequest.getHeader(CommonHttpHeader.ORIGIN)
+    val requestMethodHeader = httpServletRequest.getHeader(CommonHttpHeader.ACCESS_CONTROL_REQUEST_METHOD)
+    lazy val validMethods = getValidMethodsForResource(httpServletRequest.getRequestURI)
+
+    val requestType = (Option(origin), isOptions, Option(requestMethodHeader)) match {
+        case (Some(_), true, Some(_)) => PreflightRequest
+        case (Some(_), _, None) => ActualRequest
+        case _ => NonCorsRequest
+    }
+
+    val requestedMethod = requestType match {
+      case PreflightRequest => requestMethodHeader
+      case _ => httpServletRequest.getMethod
+    }
+
+    val validationResult = requestType match {
+      case NonCorsRequest => Pass
+      case _ =>
+        (isOriginAllowed(origin), validMethods.exists(requestedMethod == _)) match {
+          case (true, true) => Pass
+          case (false, _) => OriginNotAllowed
+          case (true, false) => MethodNotAllowed
+        }
+    }
+
+    validationResult match {
+      case Pass =>
+        requestType match {
+          case NonCorsRequest => filterChain.doFilter(httpServletRequest, httpServletResponse)
+          case PreflightRequest =>
+            httpServletResponse.setHeader(CommonHttpHeader.ACCESS_CONTROL_ALLOW_CREDENTIALS, "true")
```

Comment:
It doesn't matter in this case, but I'll note that true.toString would work as a replacement to the string literal "true". Just a "huh" type of thing.